### ISSUE TITLE
ollama modelinstance can permafail with "connection refused"

### ISSUE THREAD

**Issue by lukemarsden** [helixml/helix#242]
the runner serving ollama_mistral will just start doing this, **bringing down our production service**:
```
:x: there was a session error https://app.tryhelix.ai/session/XXX failed to get response from inference API: Post "http://localhost:41533/v1/chat/completions": dial tcp 127.0.0.1:41533: connect: connection refused
```

we need to properly run this down and fix the root cause. i've put in a temporary fix here to just exit and restart the runner when we detect this condition, but we need to fix it properly: https://github.com/helixml/helix/pull/241

sometimes, presumably the ollama server exits but we don't handle that by restarting it _or_ cleaning up our record of the model instance.

one issue is that pkg/runner/controller.go is the only call to m.Stop(), which cleans up the modelinstance from r.activeModelInstances
```
		err := m.Stop()
			if err != nil {
				log.Error().Msgf("error stopping model instance %s: %s", m.ID(), err.Error())
			}
			r.activeModelInstances.Delete(m.ID())
```

but nothing cleans up r.activeModelInstances if the ollama instance exits itself. but that might not be the root cause, the ollama instance itself keeps running even when the ollama process exits:
```
	if err := cmd.Wait(); err != nil {
			log.Error().Msgf("Ollama model instance exited with error: %s", err.Error())

			errMsg := string(stderrBuf.Bytes())
			if i.currentSession != nil {
				i.errorSession(i.currentSession, fmt.Errorf("%s from cmd - %s", err.Error(), errMsg))
			}

			return
		}

		log.Info().Msgf("🟢 Ollama model instance stopped, exit code=%d", cmd.ProcessState.ExitCode())
```

nothing here causes the goroutine later in `Start`:
```
go func() {
		for {
			select { // ...
```
to exit, or close the workCh.

so to summarize, if the ollama process exits or is killed:
* nothing stops the OllamaModelInstance or closes the channel
* nothing deletes the model instance from controller's activeModelInstances

except the sys.Exit i added in https://github.com/helixml/helix/pull/241

this issue is to properly handle this case without causing any errors to surface to the user (ideally), e.g. by restarting the ollama process if it exits or gives us "connection refused", and by not restarting the entire runner

**Comment by philwinder** [comment]
I've been seeing this again with the following setup: local dev, single luke@prime runner.

```sh
for i in `seq 1 10`; do curl --request POST \        
  --url http://localhost:8080/v1/chat/completions \
  --header 'Authorization: Bearer hl-95i-ORkQBZmUbVJHWIo04yNdylT8w3gBCLX_L3zPF7k=' \
  --header 'Content-Type: application/json' \
  --data '{
  "model": "llama3:instruct",
  "messages": [
    {
      "role": "user",
      "content": "Write a 1000 word essay on Yorkshire."
    }
  ]
}' & ; done
```

results in:
```
runner-1  | 2024-08-14T18:55:29Z ERR api/pkg/runner/llm_ollama_model_instance.go:156 > error processing request error="failed to get response from inference API: Post \"http://localhost:43851/v1/chat/completions\": dial tcp 127.0.0.1:43851: connect: connection refused" session_id=
runner-1  | 2024-08-14T18:55:29Z ERR api/pkg/runner/llm_ollama_model_instance.go:159 > detected connection refused, exiting and hoping we get restarted - see https://github.com/helixml/helix/issues/242
```

**Comment by philwinder** [comment]
Fixed in https://github.com/helixml/helix/releases/tag/0.10.2
